src/Mod/Sketcher/App/SketchObject.cpp

```diff
@@ -7215,98 +7213,90 @@ bool SketchObject::insertBSplineKnot(int GeoId, double param, int multiplicity)
+    const auto& poles{bsp->getPoles()};
+    const auto& newPoles{bspline->getPoles()};
+    std::vector<std::optional<int>> poleIndexInNew{std::size(poles)};
+    std::transform(poles.begin(), poles.end(), poleIndexInNew.begin(),
+                   [&newPoles, &findIndex](const auto& pole) { return findIndex(newPoles, pole); });
 
-    const std::vector<Sketcher::Constraint*>& cvals = Constraints.getValues();
+    const auto& knots{bsp->getKnots()};
+    const auto& newKnots{bspline->getKnots()};
+    std::vector<std::optional<int>> knotIndexInNew{std::size(knots)};
+    std::transform(knots.begin(), knots.end(), knotIndexInNew.begin(),
+                   [&newKnots, &findIndex](auto knot) { return findIndex(newKnots, knot); });
 
-    std::vector<Constraint*> newcVals(0);
+    // Handle constraints
+    std::vector<int> geometryIdsToDelete{};
+    const auto& cvals{Constraints.getValues()};
+    std::vector<Sketcher::Constraint*> newcVals{};
 
-    // modify pole and knot constraints
     for (const auto& constr : cvals) {
         if (!(constr->Type == Sketcher::InternalAlignment && constr->Second == GeoId)) {
-            newcVals.push_back(constr);
+            newcVals.emplace_back(constr);
             continue;
         }
 
-        std::vector<int>* indexInNew = nullptr;
-
-        if (constr->AlignmentType == Sketcher::BSplineControlPoint) {
-            indexInNew = &poleIndexInNew;
-        }
-        else if (constr->AlignmentType == Sketcher::BSplineKnotPoint) {
-            indexInNew = &knotIndexInNew;
-        }
-        else {
-            // it is a bspline geometry, but not a controlpoint or knot
-            newcVals.push_back(constr);
-            continue;
+        const std::vector<std::optional<int>>* indexInNew{nullptr};
+        switch (constr->AlignmentType) {
+            case Sketcher::BSplineControlPoint:
+                indexInNew = &poleIndexInNew;
+                break;
+            case Sketcher::BSplineKnotPoint:
+                indexInNew = &knotIndexInNew;
+                break;
+            default:
+                // It is a bspline geometry, but not a control point or knot
+                newcVals.emplace_back(constr);
+                continue;
         }
 
-        if (indexInNew && indexInNew->at(constr->InternalAlignmentIndex) == -1) {
-            // it is an internal alignment geometry that is no longer valid
+        if (!indexInNew->at(constr->InternalAlignmentIndex).has_value()) {
+            // It is an internal alignment geometry that is no longer valid
             // => delete it and the pole circle
-            delGeoId.push_back(constr->First);
+            geometryIdsToDelete.emplace_back(constr->First);
             continue;
         }
 
-        Constraint* newConstr = constr->clone();
-        newConstr->InternalAlignmentIndex = indexInNew->at(constr->InternalAlignmentIndex);
-        newcVals.push_back(newConstr);
+        auto* newConstr{constr->clone()};

```

Comment:
I think unless there's some kind of long qualifier, it's best to just be explicit as for the type.

Reply:
I'll see what I can do and ping you afterwards